### ISSUE TITLE
Replace legacy WP dashboard widget with refactored version in production

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2240]
Eventually the refactored version of the WP dashboard widget should be exposed in production, and the legacy version should be deleted.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The `widgets.wpDashboard` feature flag should be removed. Conditionals wrapped into a check for it in JS should be removed, i.e. the inner parts should run unconditionally.
* The `LegacyWPDashboardModules` component and all of its inner components that are not used anywhere else should be removed.
* All usages of the `googlesitekit.WPDashboardHeader` and `googlesitekit.WPDashboardModule` filters should be removed.
* All folders `assets/js/modules/*/components/wp-dashboard` should be removed.
* The new version (now available in production builds) should look the same like the legacy version before (should already be taken care of via #2075, but added here to double-check).

## Implementation Brief

* Edit `feature-flags.json` to remove references to `wpDashboard`
* Edit `/assets/js/components/wp-dashboard/WPDashboardApp.js` to remove the condition which checks if `wpDashboard` feature flag is enabled, i.e the inner parts should run unconditionally.
* Remove the following files/folders:
	- `/assets/js/components/wp-dashboard/LegacyWPDashboardModules`
	- `/assets/js/components/wp-dashboard/LegacyWPDashboardModule`
	- `/assets/js/components/wp-dashboard/LegacyWPDashboardHeader`
	- `/assets/js/modules/*/components/wp-dashboard`
* Remove `googlesitekit.WPDashboardHeader` and/or `googlesitekit.WPDashboardModule` filters in the following files:
	- `stories/wp-dashboard.stories.js`
	- `assets/js/modules/search-console/index.legacy.js`
	- `assets/js/modules/analytics/index.legacy.js`

### Test Coverage

* No changes to tests needed.

### Visual Regression Changes

* There should not be any visual regressions, though storybook has been updated, for comparison:
* **New**: [Changed data](https://google.github.io/site-kit-wp/storybook/pull/2672/?path=/story/wordpress--wordpress-dashboard)
* **Legacy**: [Different data, same look](https://google.github.io/site-kit-wp/storybook/pull/2752/?path=/story/wordpress--wordpress-dashboard)

## QA Brief

* Open **/wp-admin/index.php** in one tab with the prod build of the plugin (no feature flags enabled), with **analytics** and **search console** modules **enabled**. 
* Open the same **/wp-admin/index.php** with a previous version (e.g. the last sitekit release).
* The Sitekit dashboard widget section should look identical (see reference screenshot: https://github.com/google/site-kit-wp/blob/d7f1acfc35ea9aa5b1f4ce15f54d792ef23b75c5/tests/backstop/reference/google-site-kit_WordPress_WordPress_Dashboard_0_document_2_large.png )
 
## Changelog entry

* Remove legacy version of the WP Dashboard widget and replace it with Site Kit widget-based refactored version.

**Comment by asvinb** [comment]
@felixarntz In the AC, it's mentioned that the `widgets.wpDashboard` feature flag should be removed. However we don't have that flag. We have only `dashboard` and `pageDashboard`. I wrote the IB assuming you meant `dashboard`

**Comment by kostyalmm** [comment]
@asvinb the `wpDashboard` was added in #2075. It is still in merge review. This issue is applicable only after that

**Comment by asvinb** [comment]
Thanks @kostyalmm

**Comment by felixarntz** [comment]
@asvinb Can you please update the references in the IB to use `wpDashboard`?

**Comment by asvinb** [comment]
Updated @felixarntz

**Comment by aaemnnosttv** [comment]
@asvinb 

> All references to deleted files must be updated in other files.

Not to nitpick but would you please clarify what you mean by this? If a deleted file still has references to it then it probably shouldn't be deleted. I think it's safe to say that all `LegacyWPDashboard*` components should be deleted as part of this. If there are any exceptions, let's call them out. If not I think this point can be removed since updating leftover imports would be required for things to not break anyways.

**Comment by asvinb** [comment]
@aaemnnosttv Thanks. I meant that if the deleted files are referenced elsewhere in the app, then we should make sure we remove them from those files, just in case a stray `LegacyWPDashboard*` component is being used somewhere. I removed the last point in case to clear any confusion.

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by ivankruchkoff** [comment]
Got the test case resolved, but the blocker #2610 will need to be resolved before we can push this up.

**Comment by felixarntz** [comment]
@ivankruchkoff I'm noticing now this issue is not including anything about Storybook, where we have an existing story for the _legacy_ WP dashboard widget. Now that we remove these pieces, we need to also remove this story, so we have to ensure there is a new one that covers the refactored version of the WP dashboard widget. I think adding it to this issue is a bit late now and would increase scope, so it might be best to open a separate one. If so, we should hold off merging this one though until we also have Storybook coverage. cc @aaemnnosttv

**Comment by ivankruchkoff** [comment]
@felixarntz yep, I noticed the storybook issue (actually @eugene-manuilov did), I've already added the storybook updates (see https://github.com/google/site-kit-wp/pull/2672/files#diff-ffa8d0ded69e36d29e753e8f9ce1c5f221d7c1203bd4734d53ae598778927141R87). The only thing missing is the work @johnPhillips is doing in #2610 so I can remove the updated test screenshot and add the test for the top28 dashboard to this test scenario too, https://github.com/google/site-kit-wp/blob/c348f159e7ad70f42ba8ac5009beda5e77127181/tests/backstop/reference/google-site-kit_WordPress_WordPress_Dashboard_0_document_0_small.png

**Comment by felixarntz** [comment]
@ivankruchkoff Excellent! #2610 is now merged, so this should be good to proceed.

**Comment by cole10up** [comment]
### QA ⚠️

Checked my personal site's /index.php dashboard widget on the current release, and a dev build seeing the following:

Develop branch selecting development build:
![image](https://user-images.githubusercontent.com/57115907/108806966-010af800-7558-11eb-9707-c21ec4acda6b.png)

Latest release:
![image](https://user-images.githubusercontent.com/57115907/108806995-16802200-7558-11eb-9f70-f647b95a686b.png)

Here's the reference image:
![image](https://user-images.githubusercontent.com/57115907/108807141-78d92280-7558-11eb-8ecb-6d7f6630543d.png)

Am I missing something here @ivankruchkoff?

**Comment by ivankruchkoff** [comment]
The sitekit logo is part of the storybook story, so that can be ignored as part of this ticket.

What is important, is to figure out the slight difference in numbers, and especially figure out the difference in percentage in Total Impressions. For this, I'll get the requests and responses for the legacy request and for the new request to see if there are different dates or some different data being returned.

**Comment by ivankruchkoff** [comment]
Having compared the before and after versions of the dashboard, there is one main reason for the discrepancy, and as a result a small bug that I've pushed a fix for.

**Total Clicks / Total Impressions**
Old way: 54 rows from 2020-12-31 to 2021-02-22
New way: 55 rows from 2020-12-31 to 2021-02-23

**Total Unique Visitors**
Old: 20201231, 56 rows - between 39580 and 40531
New: 20201231, 56 rows - between 39533 and 40531

<img width="620" alt="Screen Shot 2021-02-24 at 6 57 49 PM" src="https://user-images.githubusercontent.com/373931/109105877-9b4b7700-76e3-11eb-8bd2-d22df5557231.png">
<img width="604" alt="Screen Shot 2021-02-24 at 6 57 46 PM" src="https://user-images.githubusercontent.com/373931/109105879-9be40d80-76e3-11eb-8eda-0410e42fd42b.png">

Interestingly, we're also returning report data in reverse for the two versions, but that doesn't impact the end result.
53 = 13 clicks ( 0 index in other)
52 = 5 clicks ( 1 index in other)
51 = 19 clicks (2 index in other)

<img width="367" alt="Screen Shot 2021-02-24 at 6 16 43 PM" src="https://user-images.githubusercontent.com/373931/109105956-b74f1880-76e3-11eb-887c-14bd7a5a97e1.png">

<img width="435" alt="Screen Shot 2021-02-24 at 6 16 59 PM" src="https://user-images.githubusercontent.com/373931/109105973-bddd9000-76e3-11eb-9dea-2b5a1ef4a85d.png">

One issue that I did find and pushed a branch for ( cc @felixarntz / @aaemnnosttv ):

We're not accounting for different number of days in totalImpressions vs totalOlderImpressions
https://github.com/google/site-kit-wp/blob/f165c3996ea890d22877b4ec8f13020915fdd4d7/assets/js/components/wp-dashboard/WPDashboardImpressions.js#L79


const reportArgs is:
```
{
  "startDate":"2020-12-30",
  "endDate":"2021-02-23",
  "dimensions":"date"
}
```

data.length = 55
latestData.length = 28
olderData.length = 27

So we are comparing 28 days worth of impressions with previous 27 days. (Off by one)

Proposal: Ensure that both latestData and olderData have the same amount of days being compared, so in a collection of 55 days, we'd compare the first 28 days with the last 28 days (and there's an overlap of one day).
https://github.com/google/site-kit-wp/pull/2872/files 

Alternatively, we can factor in days in the calculation for change:
```javascript
const totalImpressions = sumObjectListValue( latestData, 'impressions' );
const totalOlderImpressions = sumObjectListValue( olderData, 'impressions' );
const totalImpressionsChange = calculateChange( totalOlderImpressions, totalImpressions );
```

Becomes this:
```javascript
const totalImpressions = sumObjectListValue( latestData, 'impressions' );
const totalOlderImpressions = sumObjectListValue( olderData, 'impressions' );
const impressionsPerDay = totalImpressions / latestData.length;
const olderImpressionsPerDay = totalOlderImpressions / olderData.length;
const totalImpressionsChange = calculateChange( impressionsPerDay, olderImpressionsPerDay );
```

Also worth noting that an upgrading user would get slightly different numbers in their dashboard pre and post upgrade as in both of Cole's and my screenshots.

**Comment by aaemnnosttv** [comment]
Thanks for the thorough report @ivankruchkoff! Taking a look now.

> Interestingly, we're also returning report data in reverse for the two versions, but that doesn't impact the end result.

I think this is probably an unintended side-effect of the changes in your branch since `Array.reverse` modifies the array in place while `slice` makes a copy.

**Comment by aaemnnosttv** [comment]
@ivankruchkoff @felixarntz – the problem seems to be due to an odd number of total rows is being returned for the report, but **it should always be even** since `compare: true` should always double the number of rows, and thus always be evenly divisible by `2`, with each segment's length equal to the number of days in the range.

We're seeing `55` rows returned for a `28` day range instead of `56`. This results in the current half being `28` days and the previous being `27`. It may have been thought that the end date of the previous range and the starting date of the current range would be the same, but I don't think this should be the case since date range dates are inclusive.

Digging into this a bit more, the problem seems to be related to the change in `offsetDays`. When we request an offset of `1`, the last day is not included in the response (even though it is included in the request to the Google API) resulting in a range `1` shorter than it should be.

![image](https://user-images.githubusercontent.com/1621608/109138801-6f6fd600-7763-11eb-8c75-b460e7d1f4fc.png)

**Actual request to Google API**
| `offsetDays: 1` | `offsetDays: 2` |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/1621608/109139446-35eb9a80-7764-11eb-8c41-0730b4d6ecb8.png) | ![image](https://user-images.githubusercontent.com/1621608/109139097-c83f6e80-7763-11eb-909c-0a86ebbdb119.png) |
| ![image](https://user-images.githubusercontent.com/1621608/109139301-09d01980-7764-11eb-99ad-ed4d2e493973.png) | ![image](https://user-images.githubusercontent.com/1621608/109139113-cecde600-7763-11eb-9529-e3d17c4e6ebe.png) |
| ![image](https://user-images.githubusercontent.com/1621608/109139128-d3929a00-7763-11eb-88e4-2ccf242a6c80.png) | ![image](https://user-images.githubusercontent.com/1621608/109139295-076dbf80-7764-11eb-814d-19acf16a2241.png) |

As far as I can tell though, the client date range parsing is consistent with the server parsing so I'm not sure why there would be a difference between new and legacy.

**Comment by wpdarren** [comment]
@aaemnnosttv @ivankruchkoff I have changed the pipeline from QA to Code review, please let us know when this is ready for QA.  c.c. @cole10up

**Comment by aaemnnosttv** [comment]
After taking another look at this and comparing with the previous release using a clean cache I found that the data shown (including percentage changes) was consistent between 1.26.0 and the latest for 1.27.0.

Assigning to @wpdarren to give this another pass.

**Comment by wpdarren** [comment]
@cole10up are you okay to look at this ticket? It'll be quicker than me taking a look since you've already spent time on the QA.

**Comment by cole10up** [comment]
### QA ⚠️

@aaemnnosttv @ivankruchkoff 

I ran the same test just now with the latest zip versus the current release in an incognito window. Here are my findings.

Latest zip:
![image](https://user-images.githubusercontent.com/57115907/109260269-b0dba200-77b2-11eb-8c5b-bf03b6d70f53.png)

Latest release:
![image](https://user-images.githubusercontent.com/57115907/109260293-bfc25480-77b2-11eb-9604-86992597095e.png)

**Comment by ivankruchkoff** [comment]
Second your findings @cole10up cc @aaemnnosttv , tested now, see a difference, tested in the AM Pacific Time, saw no difference. 1.2.6 vs 1.2.7 (including 3 followup PR fixes)
<img width="623" alt="Screen Shot 2021-02-25 at 11 13 43 PM" src="https://user-images.githubusercontent.com/373931/109268184-8b08ca00-77bf-11eb-96b7-c4ec93aeecbb.png">
<img width="618" alt="Screen Shot 2021-02-25 at 11 12 13 PM" src="https://user-images.githubusercontent.com/373931/109268187-8c39f700-77bf-11eb-9ab1-5a148ed86e3f.png">

**Comment by ivankruchkoff** [comment]
Some state data from the two different ways this is calculated, i think the important thing here is the discrepancy between current and previous values between the two versions:
```json
{
  "overview":{
    "totalSessions":"43171",
    "averageBounceRate":"90.19017395937088",
    "averageSessionDuration":"24.63866947719534",
    "totalSessionsChange":0.00645777964284049,
    "averageBounceRateChange":-0.009215458225361255,
    "averageSessionDurationChange":0.05915687306748486,
    "goalCompletions":"25",
    "goalCompletionsChange":-0.10714285714285714,
    "totalPageViews":"54653",
    "totalPageViewsChange":0.021150576409259915
  },
  "totalUsers":"40529",
  "previousTotalUsers":"40711"
}
```

vs 

```json
{
  "startDate":"2021-01-28",
  "endDate":"2021-02-24",
  "compareStartDate":"2020-12-31",
  "compareEndDate":"2021-01-27",
  "metrics":[
    {
      "expression":"ga:users",
      "alias":"Total Users"
    }
  ]
}
```
and
```json
[
  {
    "nextPageToken":null,
    "columnHeader":{
      "dimensions":null,
      "metricHeader":{
        "metricHeaderEntries":[
          {
            "name":"Total Users",
            "type":"INTEGER"
          }
        ]
      }
    },
    "data":{
      "dataLastRefreshed":null,
      "isDataGolden":true,
      "rowCount":1,
      "samplesReadCounts":null,
      "samplingSpaceSizes":null,
      "rows":[
        {
          "dimensions":null,
          "metrics":[
            {
              "values":[
                "39777"
              ]
            },
            {
              "values":[
                "40531"
              ]
            }
          ]
        }
      ],
      "totals":[
        {
          "values":[
            "39777"
          ]
        },
        {
          "values":[
            "40531"
          ]
        }
      ],
      "minimums":[
        {
          "values":[
            "39777"
          ]
        },
        {
          "values":[
            "40531"
          ]
        }
      ],
      "maximums":[
        {
          "values":[
            "39777"
          ]
        },
        {
          "values":[
            "40531"
          ]
        }
      ]
    }
  }
]
```

**Comment by ivankruchkoff** [comment]
Also via dev build of https://github.com/google/site-kit-wp/pull/2672/commits/7e1f6c4a99c4785591d2b37f65b1c2183de26911 we get confirmation of the same numbers as 1.2.6 vs 1.2.7 corresponding to different ways of rendering the WP dashboard widget.
<img width="609" alt="Screen Shot 2021-02-25 at 11 39 27 PM" src="https://user-images.githubusercontent.com/373931/109270401-d1135d00-77c2-11eb-9f0f-0236c2c23d08.png">

**Comment by aaemnnosttv** [comment]
@ivankruchkoff unless you're testing with a completely session storage then I think this may be partly (or entirely) due to a change in the cache age between new and old.

The legacy `withData` wrapped components such as `WPAnalyticsDashboardWidgetOverview` are using a cache of up to 1 day long, whereas the new components will only use the default 1 hour cache TTL. So at some point, the new component will fetch fresher data than the old (unless you're emptying the request cache every time).

https://github.com/google/site-kit-wp/blob/7e1f6c4a99c4785591d2b37f65b1c2183de26911/assets/js/modules/analytics/components/wp-dashboard/WPAnalyticsDashboardWidgetOverview.js#L135

I've tested with multiple sites using 7e1f6c4 and have yet to reproduce this with a clean cache.

I'm still not sure however how the top pages table is returning 4 items instead of 5 in some cases. From what I've seen this is due to the response from the API (perhaps due to the `limit` being decreased to 5, but still not sure why that would result in _less_ than 5 returned 🤔 ).

**Comment by felixarntz** [comment]
@aaemnnosttv @ivankruchkoff @cole10up I am still unable to get any difference to show up. I double-checked the API response data that I get back between the last release and latest `master`, and the numbers match entirely.

Can you please check the following?

1. Make sure the Site Kit tester plugin is disabled, or that all feature flags are deactivated.
2. Use a production build of `master`.
3. Clean your session storage, open the "Network" tab to see API requests (XHR filter) then load the WP dashboard widget.
4. Open a new second tab for side-by-side comparison.
5. Switch to a production build of 1.26.0 release.
6. In the second tab, clean your session storage, open the "Network" tab to see API requests (XHR filter) then load the WP dashboard widget.
7. Now compare the numbers and check the source of the discrepancy if there is any:
    * If the Search Console numbers ("Impressions" and "Clicks") are different between the two versions, inspect the "searchanalytics" request (`master`) and compare it to the "searchanalytics" portion of the `data` batch request (1.26.0; in that version all requests are bundled within this one batch request). Check that the responses for these two requests return the same amount of rows, and do a high-level check that the start and end dates within these rows match. Please post any discrepancy between these two API responses here if there is any.
    * If the Analytics "Total Unique Visitors" numbers are different between the two versions, inspect the "report" request which includes `ga:users` (`ga%3Ausers`) in its query parameters (`master`) with the similar "report" request portion of the `data` batch request (1.26.0). Check that the responses for these two requests include the same total numbers for both current and previous period.
    * If the Analytics "Avg. Time on Page" numbers are different between the two versions, inspect the "report" request which includes `ga:avgSessionDuration` (`ga%3AavgSessionDuration`) in its query parameters (`master`) with the similar "report" request portion (here the report includes also other metrics in the same result set) of the `data` batch request (1.26.0). Check that the responses for these two requests include the same total `ga:avgSessionDuration` metric numbers for both current and previous period.

**Comment by ivankruchkoff** [comment]
@felixarntz try again 7pm Pacific time, or maybe 9pm. This bug is time dependent. Yesterday in the morning with @aaemnnosttv I couldn't repro, and in the evening I could. Here is how the data looks now:
<img width="612" alt="Screen Shot 2021-02-26 at 9 25 37 AM" src="https://user-images.githubusercontent.com/373931/109333770-b79b0100-7814-11eb-96c3-201befd2bd28.png">

**Comment by felixarntz** [comment]
@ivankruchkoff Ah okay, I'll check again later then!

**Comment by aaemnnosttv** [comment]
@felixarntz I thought maybe this had something to do with the legacy `withData` components using a 1 day TTL (`maxAge`) on their requests, but @cole10up showed it to me earlier today, even resetting Site Kit in between so it's definitely time related with the API responses.

**Comment by felixarntz** [comment]
@aaemnnosttv @ivankruchkoff @wpdarren Tested this again around 8pm PT (2021-02-27), now indeed the numbers are different. This appears to happen for the following reason:
* For both Analytics and Search Console, `master` returns data from `2021-01-02` until `2021-01-26` (`2021-01-30` until `2021-02-26` for main date range, `2021-01-02` until `2021-01-29` for compare date range).
* For both Analytics and Search Console, 1.26.0 returns data from `2021-01-03` until `2021-01-27` (`2021-01-31` until `2021-02-27` for main date range, `2021-01-03` until `2021-01-30` for compare date range). Because this goes all the way until the current day, it for some reason ignores the expected 1 day offset (for the current date there is never any data).

This could be because the date parsing happens on the server for the legacy data API, thus relying on the timezone of the server, which might cause this slightly different behavior, where now the date range is parsed on the client. Anyway, the new numbers returned by `master` are the correct ones, so we can proceed with this as is.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

Looking at QAB, it suggests that the dashboard widget section should look identical, but as discussed at length, I can see that the numbers are different. Reading the last comment by @felixarntz I am assuming the master figures are correct and the difference is to be expected due to API/data.  

**Master**
[Screenshot](https://user-images.githubusercontent.com/73545194/109435433-6cbbfc00-7a55-11eb-874f-95dd264cc324.png)

**Version 1.26.0**
[Screenshot](https://user-images.githubusercontent.com/73545194/109435453-7e050880-7a55-11eb-9a9b-6092845d90cb.png)

c.c. @cole10up @aaemnnosttv I've not looked at this ticket before so just wanting to make sure we are okay moving this on based on Felix's latest comment.
